x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/persistence/BatchedDocumentsIterator.java

```diff
@@ -85,8 +85,12 @@ public Deque<T> next() {
             SearchScrollRequest searchScrollRequest = new SearchScrollRequest(scrollId).scroll(CONTEXT_ALIVE_DURATION);
             searchResponse = client.searchScroll(searchScrollRequest).actionGet();
         }
-        scrollId = searchResponse.getScrollId();
-        return mapHits(searchResponse);
+        try {
+            scrollId = searchResponse.getScrollId();
+            return mapHits(searchResponse);
```

Comment:
This is good, the `mapHits` always creates new objects from what I can tell. I would be scared if it kept a reference to a search hit.

cc @droberts195 ^ from what I read, this is ok.

Reply:
Yes, I think this is OK. The `_source` of the hits gets parsed by the appropriate result parser synchronously. Then a list of ML results objects gets returned. So by the time the method returns the search hits are no longer needed or referenced.